src/parcsr_ls/par_cheby_device.c

```diff
@@ -0,0 +1,176 @@
+ * @param[in] scale Whether or not to scale by diagonal
+ * @param[in] scale Whether or not to use a variant
+ * @param[in,out] u Initial/updated approximation
+ * @param[out] v Temp vector
+ * @param[out] v Temp Vector
+ */
+HYPRE_Int
+hypre_ParCSRRelax_Cheby_SolveDevice(hypre_ParCSRMatrix *A, /* matrix to relax with */
+                                    hypre_ParVector    *f, /* right-hand side */
+                                    HYPRE_Real         *ds_data,
+                                    HYPRE_Real         *coefs,
+                                    HYPRE_Int           order, /* polynomial order */
+                                    HYPRE_Int           scale, /* scale by diagonal?*/
+                                    HYPRE_Int           variant,
+                                    hypre_ParVector    *u, /* initial/updated approximation */
+                                    hypre_ParVector    *v /* temporary vector */,
+                                    hypre_ParVector    *r /*another temp vector */)
+{
+   hypre_CSRMatrix *A_diag = hypre_ParCSRMatrixDiag(A);
+   HYPRE_Real      *u_data = hypre_VectorData(hypre_ParVectorLocalVector(u));
+   HYPRE_Real      *f_data = hypre_VectorData(hypre_ParVectorLocalVector(f));
+   HYPRE_Real      *v_data = hypre_VectorData(hypre_ParVectorLocalVector(v));
+
+   HYPRE_Real *r_data = hypre_VectorData(hypre_ParVectorLocalVector(r));
+
+   HYPRE_Int i;
+   HYPRE_Int num_rows = hypre_CSRMatrixNumRows(A_diag);
+
+   HYPRE_Real  mult;
+   HYPRE_Real *orig_u;
+
+   HYPRE_Int cheby_order;
+
+   HYPRE_Real *tmp_data;
+
+   hypre_ParVector *tmp_vec;
+
+   /* u = u + p(A)r */
+
+   if (order > 4) order = 4;
+   if (order < 1) order = 1;
+
+   /* we are using the order of p(A) */
+   cheby_order = order - 1;
+
+   orig_u = hypre_CTAlloc(HYPRE_Real, num_rows, hypre_CSRMatrixMemoryLocation(A_diag));
+
+   if (!scale)
+   {
+      /* get residual: r = f - A*u */
+      hypre_ParVectorCopy(f, r);
+      hypre_ParCSRMatrixMatvec(-1.0, A, u, 1.0, r);
+
+      /* o = u; u = r .* coef */
+      HYPRE_THRUST_CALL(
+          for_each,
+          thrust::make_zip_iterator(thrust::make_tuple(orig_u, u_data, r_data)),
+          thrust::make_zip_iterator(thrust::make_tuple(orig_u + num_rows, u_data + num_rows, r_data + num_rows)),
+          save_and_scale<HYPRE_Real>(coefs[cheby_order]));
+
+      for (i = cheby_order - 1; i >= 0; i--)
+      {
+         hypre_ParCSRMatrixMatvec(1.0, A, u, 0.0, v);
+         mult = coefs[i];
+         /* u = mult * r + v */
+
+         HYPRE_THRUST_CALL(
+             for_each,
+             thrust::make_zip_iterator(thrust::make_tuple(u_data, r_data, v_data)),
+             thrust::make_zip_iterator(thrust::make_tuple(u_data + num_rows, r_data + num_rows, v_data + num_rows)),
+             oop_axpy<HYPRE_Real>(mult));
+      }
+
+      /* u = o + u */
+      HYPRE_THRUST_CALL(transform, orig_u, orig_u + num_rows, u_data, u_data, xpy<HYPRE_Real>());
```

Comment:
you can do `HYPRE_THRUST_CALL(transform, ..., _1 + _2)` without `xpy`

Reply:
Done (https://github.com/hypre-space/hypre/pull/388/commits/d9ab86dd14bcfc1e2e2659d4073bad5c2d37195d)